New gcc breaks travis build with -Werror=address-of-packed-member

<!--
PLEASE HELP US PROCESS GITHUB ISSUES FASTER BY PROVIDING THE FOLLOWING INFORMATION.
-->

## I'm submitting a…
<!-- Please check one of the following options with "x" -->
<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

i3 4.17

Travis build on next: https://travis-ci.org/i3/i3/jobs/579505745
```
  CC       src/i3-config_parser.o

../../i3/src/randr.c: In function 'randr_query_outputs_15':

../../i3/src/randr.c:667:33: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  667 |             update_if_necessary(&(new->rect.x), monitor_info->x) |

      |                                 ^~~~~~~~~~~~~~

../../i3/src/randr.c:668:33: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  668 |             update_if_necessary(&(new->rect.y), monitor_info->y) |

      |                                 ^~~~~~~~~~~~~~

../../i3/src/randr.c:669:33: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  669 |             update_if_necessary(&(new->rect.width), monitor_info->width) |

      |                                 ^~~~~~~~~~~~~~~~~~

../../i3/src/randr.c:670:33: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  670 |             update_if_necessary(&(new->rect.height), monitor_info->height);

      |                                 ^~~~~~~~~~~~~~~~~~~

../../i3/src/randr.c: In function 'handle_output':

../../i3/src/randr.c:744:40: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  744 |     bool updated = update_if_necessary(&(new->rect.x), crtc->x) |

      |                                        ^~~~~~~~~~~~~~

../../i3/src/randr.c:745:40: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  745 |                    update_if_necessary(&(new->rect.y), crtc->y) |

      |                                        ^~~~~~~~~~~~~~

../../i3/src/randr.c:746:40: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  746 |                    update_if_necessary(&(new->rect.width), crtc->width) |

      |                                        ^~~~~~~~~~~~~~~~~~

../../i3/src/randr.c:747:40: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  747 |                    update_if_necessary(&(new->rect.height), crtc->height);

      |                                        ^~~~~~~~~~~~~~~~~~~

../../i3/src/randr.c: In function 'randr_query_outputs':

../../i3/src/randr.c:874:37: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  874 |             if (update_if_necessary(&(output->rect.width), width) |

      |                                     ^~~~~~~~~~~~~~~~~~~~~

../../i3/src/randr.c:875:37: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  875 |                 update_if_necessary(&(output->rect.height), height))

      |                                     ^~~~~~~~~~~~~~~~~~~~~~

../../i3/src/randr.c:878:33: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  878 |             update_if_necessary(&(other->rect.width), width);

      |                                 ^~~~~~~~~~~~~~~~~~~~

../../i3/src/randr.c:879:33: error: taking address of packed member of 'struct Rect' may result in an unaligned pointer value [-Werror=address-of-packed-member]

  879 |             update_if_necessary(&(other->rect.height), height);

      |                                 ^~~~~~~~~~~~~~~~~~~~~

  AR       libi3.a
```

It complains about `Rect` being packed:
https://github.com/i3/i3/blob/225ebb0d588e57d7cdb4d8bd269f7f0b791c9368/include/data.h#L157-L162

@stapelberg do we need this or can we safely remove `__attribute__((packed))`?

--- a/include/data.h
+++ b/include/data.h
@@ -159,7 +159,7 @@ struct Rect {
     uint32_t y;
     uint32_t width;
     uint32_t height;
-} __attribute__((packed));
+};
 
 /**
  * Stores the reserved pixels on each screen edge read from a

--- a/include/util.h
+++ b/include/util.h
@@ -64,6 +64,7 @@ int max(int a, int b);
 bool rect_contains(Rect rect, uint32_t x, uint32_t y);
 Rect rect_add(Rect a, Rect b);
 Rect rect_sub(Rect a, Rect b);
+bool rect_equals(Rect a, Rect b);
 
 /**
  * Returns true if the name consists of only digits.

--- a/src/commands.c
+++ b/src/commands.c
@@ -469,7 +469,7 @@ static void cmd_resize_floating(I3_CMD, const char *way, const char *direction_s
 
     /* Did we actually resize anything or did the size constraints prevent us?
      * If we could not resize, exit now to not move the window. */
-    if (memcmp(&old_rect, &(floating_con->rect), sizeof(Rect)) == 0) {
+    if (rect_equals(old_rect, floating_con->rect)) {
         return;
     }
 

--- a/src/floating.c
+++ b/src/floating.c
@@ -322,11 +322,10 @@ void floating_enable(Con *con, bool automatic) {
 
     DLOG("Original rect: (%d, %d) with %d x %d\n", con->rect.x, con->rect.y, con->rect.width, con->rect.height);
     DLOG("Geometry = (%d, %d) with %d x %d\n", con->geometry.x, con->geometry.y, con->geometry.width, con->geometry.height);
-    Rect zero = {0, 0, 0, 0};
     nc->rect = con->geometry;
     /* If the geometry was not set (split containers), we need to determine a
      * sensible one by combining the geometry of all children */
-    if (memcmp(&(nc->rect), &zero, sizeof(Rect)) == 0) {
+    if (rect_equals(nc->rect, (Rect){0, 0, 0, 0})) {
         DLOG("Geometry not set, combining children\n");
         Con *child;
         TAILQ_FOREACH(child, &(con->nodes_head), nodes) {

--- a/src/load_layout.c
+++ b/src/load_layout.c
@@ -141,8 +141,7 @@ static int json_end_map(void *ctx) {
             // Also set a size if none was supplied, otherwise the placeholder
             // window cannot be created as X11 requests with width=0 or
             // height=0 are invalid.
-            const Rect zero = {0, 0, 0, 0};
-            if (memcmp(&(json_node->rect), &zero, sizeof(Rect)) == 0) {
+            if (rect_equals(json_node->rect, (Rect){0, 0, 0, 0})) {
                 DLOG("Geometry not set, combining children\n");
                 Con *child;
                 TAILQ_FOREACH(child, &(json_node->nodes_head), nodes) {

--- a/src/scratchpad.c
+++ b/src/scratchpad.c
@@ -279,7 +279,7 @@ void scratchpad_fix_resolution(void) {
 
     Rect new_rect = __i3_output->rect;
 
-    if (memcmp(&old_rect, &new_rect, sizeof(Rect)) == 0) {
+    if (rect_equals(new_rect, old_rect)) {
         DLOG("Scratchpad size unchanged.\n");
         return;
     }

--- a/src/util.c
+++ b/src/util.c
@@ -53,6 +53,10 @@ Rect rect_sub(Rect a, Rect b) {
                   a.height - b.height};
 }
 
+bool rect_equals(Rect a, Rect b) {
+    return a.x == b.x && a.y == b.y && a.width == b.width && a.height == b.height;
+}
+
 /*
  * Returns true if the name consists of only digits.
  *

--- a/src/x.c
+++ b/src/x.c
@@ -251,8 +251,7 @@ void x_move_win(Con *src, Con *dest) {
     state_dest->con = state_src->con;
     state_src->con = NULL;
 
-    Rect zero = {0, 0, 0, 0};
-    if (memcmp(&(state_dest->window_rect), &(zero), sizeof(Rect)) == 0) {
+    if (rect_equals(state_dest->window_rect, (Rect){0, 0, 0, 0})) {
         memcpy(&(state_dest->window_rect), &(state_src->window_rect), sizeof(Rect));
         DLOG("COPYING RECT\n");
     }
@@ -929,7 +928,7 @@ void x_push_node(Con *con) {
     bool fake_notify = false;
     /* Set new position if rect changed (and if height > 0) or if the pixmap
      * needs to be recreated */
-    if ((is_pixmap_needed && con->frame_buffer.id == XCB_NONE) || (memcmp(&(state->rect), &rect, sizeof(Rect)) != 0 &&
+    if ((is_pixmap_needed && con->frame_buffer.id == XCB_NONE) || (!rect_equals(state->rect, rect) &&
                                                                    rect.height > 0)) {
         /* We first create the new pixmap, then render to it, set it as the
          * background and only afterwards change the window size. This reduces
@@ -1008,7 +1007,7 @@ void x_push_node(Con *con) {
 
     /* dito, but for child windows */
     if (con->window != NULL &&
-        memcmp(&(state->window_rect), &(con->window_rect), sizeof(Rect)) != 0) {
+        !rect_equals(state->window_rect, con->window_rect)) {
         DLOG("setting window rect (%d, %d, %d, %d)\n",
              con->window_rect.x, con->window_rect.y, con->window_rect.width, con->window_rect.height);
         xcb_set_window_rect(conn, con->window->id, con->window_rect);

